File: backend/app/api/production_release.py
"""Explicit production release endpoint (Phase 4 — GUI_DESIGN §11.4).

HTTP surface for the "Release to production…" action carried by each
saved demand row. The route stays thin orchestration: the request
schema validates shape only (``extra="forbid"``), the Application
command owns every rule and the one-submission-one-transaction
protocol (SLICE1_DATA_MODEL §8, §13–§14), and the central handlers in
``app.api.errors`` translate typed failures — including the
confirmation-required response that carries the existing active
distribution.

Deliberate surface decisions:

- The endpoint is addressed through the initiating WorkOrderDemand
  (``/work-orders/{id}/demands/{id}/release``): the UI always releases
  from a saved demand row, and the demand context is a required part
  of the request fingerprint. This is informational release context
  only — no Movement→Demand foreign key exists and WorkOrderDemand
  never owns Movement.
- ``device_event_id`` is the client-generated UUID idempotency key,
  reused verbatim on transport retries: a replay of the committed
  release returns the original result with 200 (a fresh release is
  201) and creates nothing; reuse with a different normalized request
  is an explicit 409 that creates nothing.
- ``route_mode`` is required — the release flow explicitly confirms
  it (FLOATING by default in the UI, never defaulted here);
  ``route_template_id`` travels only with a PLANNED release.
- No client-writable actor exists on this surface (same rule as the
  intake endpoints): the informational actor context stays absent
  until an authenticated identity exists (Phase 14).
"""

import datetime

from fastapi import APIRouter, Response
from pydantic import BaseModel, ConfigDict, StrictBool, StrictInt

from app.api.dependencies import SessionDep
from app.application import production_release
from app.application.production_release import ProductionRelease
from app.domain.enums import RouteMode

router = APIRouter(prefix="/api")


class ProductionReleaseRequest(BaseModel):
    """The confirmed release submission (GUI_DESIGN §11.4)."""

    model_config = ConfigDict(extra="forbid")

    part_number: str
    # Strict: a quantity is an integer, never a coerced bool/float/text.
    quantity: StrictInt
    route_mode: RouteMode
    # Required for PLANNED, forbidden for FLOATING.
    route_template_id: int | None = None
    starting_area_id: int
    operation_id: int
    # Explicit active-quantity confirmation (SLICE1 §8.2): set by the
    # UI only after showing the existing distribution.
    confirm_active_quantity: StrictBool = False
    device_event_id: str


class ProductionReleaseResponse(BaseModel):
    """The committed release result (SLICE1 §8.6)."""

    quantity_flow_id: int
    part_number: str
    quantity: int
    route_mode: RouteMode
    # The AssignedRoute snapshot id for a PLANNED release, null for
    # FLOATING.
    assigned_route_id: int | None
    current_area_id: int
    operation_id: int
    movement_id: int
    device_event_id: str
    occurred_at: datetime.datetime


def _response(result: ProductionRelease) -> ProductionReleaseResponse:
    flow = result.flow
    movement = result.movement
    return ProductionReleaseResponse(
        quantity_flow_id=flow.id,
        part_number=flow.part_number,
        quantity=flow.quantity,
        route_mode=RouteMode(flow.route_mode),
        assigned_route_id=flow.assigned_route_id,
        current_area_id=flow.current_area_id,
        operation_id=movement.operation_id,
        movement_id=movement.id,
        device_event_id=movement.device_event_id,
        occurred_at=movement.occurred_at,
    )


@router.post("/work-orders/{work_order_id}/demands/{demand_id}/release")
def release_to_production(
    work_order_id: int,
    demand_id: int,
    body: ProductionReleaseRequest,
    session: SessionDep,
    response: Response,
) -> ProductionReleaseResponse:
    result = production_release.release_to_production(
        session,
        work_order_id=work_order_id,
        work_order_demand_id=demand_id,
        part_number=body.part_number,
        quantity=body.quantity,
        route_mode=body.route_mode,
        route_template_id=body.route_template_id,
        starting_area_id=body.starting_area_id,
        operation_id=body.operation_id,
        confirm_active_quantity=body.confirm_active_quantity,
        device_event_id=body.device_event_id,
    )
    response.status_code = 201 if result.created else 200
    return _response(result)

File: frontend/src/api/production-release.ts
// Explicit production release API (Phase 4 — GUI_DESIGN §11.4).
//
// The one write that creates production quantity. It is addressed
// through the initiating WorkOrderDemand (the UI always releases from
// a saved demand row) and protected by a client-generated
// `device_event_id` idempotency key: ONE submission keeps ONE key
// through every transport retry — a replay of the committed release
// returns the original result (`created: false`) and creates nothing.
// A new submission (a reopened dialog, a release after a completed
// one) generates a fresh key.
//
// The backend's confirmation-required outcome (the PN already has
// active quantity, SLICE1 §8.2) arrives as a 409 whose body carries
// the existing active distribution; `activeQuantityConfirmation`
// extracts it so the dialog can show the distribution and require
// explicit confirmation — resubmitting with `confirmActiveQuantity`
// and the SAME `device_event_id` (nothing was created by the 409).
//
// Production-safe: no mock data, no framework imports.

import { ApiError, apiRequestWithStatus } from './client';

export interface ProductionReleaseInput {
  partNumber: string;
  quantity: number;
  routeMode: 'FLOATING' | 'PLANNED';
  /** Required for PLANNED, absent for FLOATING. */
  routeTemplateId: number | null;
  startingAreaId: number;
  operationId: number;
  /** Set only after the existing distribution was shown (§11.4). */
  confirmActiveQuantity: boolean;
  /** The submission's idempotency key — reused verbatim on retries. */
  deviceEventId: string;
}

export interface ProductionReleaseResult {
  /** True for a fresh 201 commit, false for an idempotent replay. */
  created: boolean;
  quantityFlowId: number;
  partNumber: string;
  quantity: number;
  routeMode: 'FLOATING' | 'PLANNED';
  /** The AssignedRoute snapshot id (PLANNED), null for FLOATING. */
  assignedRouteId: number | null;
  startingAreaId: number;
  operationId: number;
  /** The appended RECEIVED Movement. */
  movementId: number;
  deviceEventId: string;
  occurredAt: string;
}

/** One existing ACTIVE flow of the PN, as shown for confirmation. */
export interface ActiveQuantityEntry {
  quantityFlowId: number;
  quantity: number;
  routeMode: 'FLOATING' | 'PLANNED';
  currentAreaId: number;
  currentAreaName: string;
}

interface ProductionReleaseWire {
  quantity_flow_id: number;
  part_number: string;
  quantity: number;
  route_mode: 'FLOATING' | 'PLANNED';
  assigned_route_id: number | null;
  starting_area_id: number;
  operation_id: number;
  movement_id: number;
  device_event_id: string;
  occurred_at: string;
}

/** A fresh UUID idempotency key for one release submission. */
export function newDeviceEventId(): string {
  return crypto.randomUUID();
}

export async function releaseToProduction(
  workOrderId: number,
  demandId: number,
  input: ProductionReleaseInput,
): Promise<ProductionReleaseResult> {
  const { status, data: wire } =
    await apiRequestWithStatus<ProductionReleaseWire>(
      `/api/work-orders/${workOrderId}/demands/${demandId}/release`,
      {
        method: 'POST',
        body: {
          part_number: input.partNumber,
          quantity: input.quantity,
          route_mode: input.routeMode,
          route_template_id: input.routeTemplateId,
          starting_area_id: input.startingAreaId,
          operation_id: input.operationId,
          confirm_active_quantity: input.confirmActiveQuantity,
          device_event_id: input.deviceEventId,
        },
      },
    );
  return {
    created: status === 201,
    quantityFlowId: wire.quantity_flow_id,
    partNumber: wire.part_number,
    quantity: wire.quantity,
    routeMode: wire.route_mode,
    assignedRouteId: wire.assigned_route_id,
    startingAreaId: wire.starting_area_id,
    operationId: wire.operation_id,
    movementId: wire.movement_id,
    deviceEventId: wire.device_event_id,
    occurredAt: wire.occurred_at,
  };
}

/**
 * The active-distribution confirmation payload of a failed release,
 * or null when the error is anything else.
 */
export function activeQuantityConfirmation(
  error: unknown,
): ActiveQuantityEntry[] | null {
  if (!(error instanceof ApiError) || error.status !== 409) return null;
  const body = error.body;
  if (!body || typeof body !== 'object') return null;
  const record = body as {
    confirmation_required?: unknown;
    existing_active_quantity?: unknown;
  };
  if (record.confirmation_required !== true) return null;
  if (!Array.isArray(record.existing_active_quantity)) return null;
  return record.existing_active_quantity.map((entry) => {
    const wire = entry as {
      quantity_flow_id: number;
      quantity: number;
      route_mode: 'FLOATING' | 'PLANNED';
      current_area_id: number;
      current_area_name: string;
    };
    return {
      quantityFlowId: wire.quantity_flow_id,
      quantity: wire.quantity,
      routeMode: wire.route_mode,
      currentAreaId: wire.current_area_id,
      currentAreaName: wire.current_area_name,
    };
  });
}
